feat: QueryCursor iterators

bennypowers opened the pull request:
Closes #30

Adds `AllMatches` and `AllCaptures` iterator methods to `QueryCursor`, following Go's [iter naming conventions](https://pkg.go.dev/iter#hdr-Naming_Conventions) as suggested in #30.

- `AllMatches` returns `iter.Seq[*QueryMatch]`
- `AllCaptures` returns `iter.Seq2[*QueryMatch, uint]`

The existing `Matches`/`Captures` methods are unchanged.

bennypowers committed:
feat: QueryCursor iterators

Closes #30

bennypowers commented:
I'm not sure about the tests here, there's probably a better way to test IterCaptures

bennypowers committed:
test: improved querycursor iterator test

topi314 commented on query.go:
```diff
@@ -767,6 +768,27 @@ func (qc *QueryCursor) Matches(query *Query, node *Node, text []byte) QueryMatch
 	return qm
 }
 
+// Iterator yielding all of the matches in the order that they were found.
+//
+// Each match contains the index of the pattern that matched, and a list of
+// captures. Because multiple patterns can match the same set of nodes,
+// one match may contain captures that appear *before* some of the
+// captures from a previous match.
```
you should also include the notes from here: https://github.com/tree-sitter/go-tree-sitter/blob/adc13ffd8b2c0b01b878fda9f7c422ce0df5fad3/query.go#L1017

bennypowers replied:
PTAL. names changed; docs added

bennypowers committed:
refactor: docs and names

bennypowers commented:
I tried something similar in a project i'm working on and got some super weird errors

```go
func allMatches(qc *ts.QueryCursor, q *ts.Query, node *ts.Node, text []byte) iter.Seq2[*ts.QueryMatch, *ts.Query] {
	qm := qc.Matches(q, node, text)
	return func(yield func (qm *ts.QueryMatch, q *ts.Query) bool) {
		for {
			m := qm.Next()
			if m == nil {
				break
			}
			if !yield(m, q) {
				return
			}
		}
	}
}
```
in the yield call i got a panic:

```
1. Thread stopped due to exception ()
   Description: Error getting throw reason: could not find symbol value for s
   Stack trace:
   Stack:
   	3  0x00007faf6f3cd11c in C.__pthread_kill_implementation
   	    at /usr/src/debug/glibc-2.41-5.fc42.x86_64/nptl/pthread_kill.c:43
   	4  0x00007faf6f373afe in C.raise
   	    at ../sysdeps/posix/raise.c:26
   	5  0x00007faf6f35b6d0 in C.abort
   	    at /usr/src/debug/glibc-2.41-5.fc42.x86_64/stdlib/abort.c:73
   	6  0x00007faf6f35b639 in C.__assert_perror_fail
   	    at /usr/src/debug/glibc-2.41-5.fc42.x86_64/assert/assert-perr.c:31
```

bennypowers commented:
Ok having played around with this, it was pebcak. I believe this code does what it says it does, I was just not properly managing the treesitter c object lifetimes

bennypowers committed:
test: full coverage for AllMatches and AllCaptures iterators

Fix tests to use the correct method names (AllMatches/AllCaptures
instead of IterMatches/IterCaptures) per Go iter naming conventions.
Add tests for early break, empty results, and multiple patterns.

Assisted-By: Claude Opus 4.6 (1M context) <noreply@anthropic.com>
